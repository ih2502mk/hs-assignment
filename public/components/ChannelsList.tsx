import React, { FC } from "react";
import Channel from "../../shared/Channel";
import styled from 'styled-components';

const List = styled.ul`
    padding: 12px;
`;

interface ChannelsListProps {
    channels: Channel[];
    onSelectChannel: (id: Channel["id"]) => void;
    className?: string;
}

export const ChannelsList: FC<ChannelsListProps> = ({
    channels,
    onSelectChannel,
    className
}) => {
    return (
        <List className={className}>
            {channels.map(({ name, id }) => (
                <li key={id}>
                    <button onClick={() => onSelectChannel(id)}>{name}</button>
                </li>
            ))}
        </List>
    );
};
