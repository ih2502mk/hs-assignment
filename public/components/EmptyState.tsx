import React, { FC, ReactNode } from "react";
import styled from "styled-components";

const Wrapper = styled.div`
    color: lightgray;
    text-align: center;
    width: 100%;
`;

export const EmptyState: FC<ReactNode> = ({ children }) => {
    return <Wrapper>{children}</Wrapper>;
};
