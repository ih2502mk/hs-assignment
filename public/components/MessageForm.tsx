import React, { FC } from "react";
import { Formik, Form, Field, ErrorMessage } from 'formik';
import Message from "../../shared/Message";

interface MessageFormProps {
    message: Message;
}

export const MessageForm: FC<MessageFormProps> = ({ message }) => {
    return (
        <Formik
            initialValues={{content: message.content}}
            onSubmit={(values, { setSubmitting }) => {
                console.log(values);
                setTimeout(() => setSubmitting(false), 100)
            }}
        >
            <Form>
                <Field as="textarea" name="content" />
                <button type="submit">Send</button>
            </Form>
        </Formik>
    );
};
