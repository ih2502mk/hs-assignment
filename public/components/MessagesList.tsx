import React, { FC, useContext, useEffect, useRef, useState } from "react";
import styled from "styled-components";
import Message from "../../shared/Message";
import { UserContext } from "../context";

const df = new Intl.DateTimeFormat("en-US", {
    weekday: "short",
    year: "numeric",
    month: "short",
    day: "numeric",
    hour: "numeric",
    minute: "numeric"
});

const List = styled.ul`
    height: calc(100vh - 300px);
    overflow-y: scroll;
`;

const MessageEntry = styled.li`
    margin: 0.6rem 0;
    border: 3px solid #eee;
    border-radius: 6px;
    padding: 0.72rem;

    &.active {
        background-color: lightyellow;
    }
`;

const MessageInfoSection = styled.div`
    font-size: 0.7rem;
    color: grey;
    padding: 0 0 0.3rem 0;
    display: flex;
    align-items: baseline;

    b {
        font-weight: bold;
        margin-right: auto;
    }

    span {
        margin-left: 0.6rem;
    }
`;

const MessageActionBtn = styled.button`
    border: none;
    background: none;
    font-size: 0.7rem;
    color: rgb(117, 190, 255);
    text-decoration: underline;
    margin: 0 0.3rem 0 0;
    width: 32px;

    :hover {
        cursor: pointer;
    }
`;

interface MessagesListProps {
    messages: Message[];
    activeMessage: Message;
    onSelectMessage: (message: Message) => void;
}

export const MessagesList: FC<MessagesListProps> = ({
    messages,
    activeMessage,
    onSelectMessage
}) => {
    const user = useContext(UserContext);
    const listEl = useRef(null);

    useEffect(() => {
        listEl.current.scrollTo(
            0,
            listEl.current.scrollHeight - listEl.current.clientHeight
        );
    }, [messages]);

    return (
        <List ref={listEl}>
            {messages.map(m => (
                <MessageEntry
                    key={m.id}
                    className={
                        activeMessage && m.id === activeMessage.id
                            ? "active"
                            : ""
                    }
                >
                    <MessageInfoSection>
                        <b>{m.author}</b>
                        {user.name === m.author ? (
                            <>
                                <MessageActionBtn
                                    onClick={() => onSelectMessage(m)}
                                >
                                    Edit
                                </MessageActionBtn>
                                <MessageActionBtn>Delete</MessageActionBtn>
                            </>
                        ) : null}
                        <span>on {df.format(m.created)}</span>
                    </MessageInfoSection>
                    <div>{m.content}</div>
                </MessageEntry>
            ))}
        </List>
    );
};
