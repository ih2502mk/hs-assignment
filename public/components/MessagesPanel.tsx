import React, { FC, useState, useEffect, useReducer } from "react";
import styled from "styled-components";
import Message from "../../shared/Message";
import Channel from "../../shared/Channel";
import { MessagesList } from "./MessagesList";
import { MessageForm } from "./MessageForm";
import { messagesReducer, actionType } from "../store/messagesReducer";
import { messagesApi } from "../api/messagesApi";

const Wrapper = styled.div``;

interface MessagesPanelProps {
    channelId: Channel["id"];
    className?: string;
}

export const MessagesPanel: FC<MessagesPanelProps> = ({
    channelId,
    className
}) => {
    const [editedMessage, setEditedMessage] = useState<Message>({
        id: null,
        content: "",
        channelId,
        author: ""
    });
    const [messages, updateMessages] = useReducer(messagesReducer, []);

    useEffect(() => {
        if (channelId) {
            messagesApi
                .getAllMessagesInChannel(channelId)
                .then(handleGetMessages);
        }
    }, [channelId]);

    const handleGetMessages = (messagesResp: { messages: Message[] }) => {
        updateMessages({
            type: actionType.init,
            payload: messagesResp.messages
        });
    };

    const handleSelectMessageForEditing = (message: Message) => {
        setEditedMessage(message);
    };

    return (
        <Wrapper className={className}>
            <MessagesList
                messages={messages}
                onSelectMessage={handleSelectMessageForEditing}
            />
            <MessageForm message={editedMessage} />
        </Wrapper>
    );
};
