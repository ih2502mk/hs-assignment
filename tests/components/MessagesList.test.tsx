import React from "react";
import renderer from "react-test-renderer";
import { MessagesList } from "../../public/components/MessagesList";
import Message from "../../shared/Message";

const testDate = new Date(Date.UTC(2020, 11, 9, 10, 10, 10, 10));

const mockMessages: Message[] = [
    { id: "1", content: "Test Msg 1", author: "John", channelId: "ch1", created: testDate},
    { id: "2", content: "Test Msg 2", author: "Bob", channelId: "ch1", created: testDate},
    { id: "3", content: "Test Msg 3", author: "Alice", channelId: "ch1", created: testDate},
    { id: "4", content: "Test Msg 4", author: "Walter", channelId: "ch1", created: testDate},
];

describe("snapshots", () => {
    it("renders a list of messages", () => {
        const component = renderer
            .create(
                <MessagesList
                    messages={mockMessages}
                    activeMessage={mockMessages[0]}
                    onSelectMessage={() => {}}
                    onDeleteMessage={() => {}}
                />
            )
            .toJSON();

        expect(component).toMatchSnapshot();
    });
});
