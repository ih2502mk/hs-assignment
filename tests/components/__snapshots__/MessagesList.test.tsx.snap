// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots renders a list of messages 1`] = `
<ul
  className="sc-bdfBwQ jUKHGV"
>
  <li
    className="sc-gsTCUz cEgtDK active"
  >
    <div
      className="sc-dlfnbm jzVabE"
    >
      <b>
        John
      </b>
      <span>
        on 
        Mon, Nov 9, 2020, 10:45 AM
      </span>
    </div>
    <div>
      Test Msg 1
    </div>
  </li>
  <li
    className="sc-gsTCUz cEgtDK"
  >
    <div
      className="sc-dlfnbm jzVabE"
    >
      <b>
        Bob
      </b>
      <span>
        on 
        Mon, Nov 9, 2020, 10:45 AM
      </span>
    </div>
    <div>
      Test Msg 2
    </div>
  </li>
  <li
    className="sc-gsTCUz cEgtDK"
  >
    <div
      className="sc-dlfnbm jzVabE"
    >
      <b>
        Alice
      </b>
      <span>
        on 
        Mon, Nov 9, 2020, 10:45 AM
      </span>
    </div>
    <div>
      Test Msg 3
    </div>
  </li>
  <li
    className="sc-gsTCUz cEgtDK"
  >
    <div
      className="sc-dlfnbm jzVabE"
    >
      <b>
        Walter
      </b>
      <span>
        on 
        Mon, Nov 9, 2020, 10:45 AM
      </span>
    </div>
    <div>
      Test Msg 4
    </div>
  </li>
</ul>
`;
